s2p/triangulation.py

```diff
@@ -255,37 +272,52 @@ def height_map_to_point_cloud(cloud, heights, rpc, H=None, crop_colorized='',
+            we are dealing with in the pixel coordinates of the original full
+            size image
         crop_colorized (optional, default ''): path to a colorized crop of a
             Pleiades image
-        off_{x,y} (optional, default None): coordinates of the point we want to
-            use as origin in the local coordinate system of the computed cloud
-        ascii_ply (optional, default false): boolean flag to tell if the output
-            ply file should be encoded in plain text (ascii).
-        utm_zone (optional, default None):
     """
-    # write rpc coefficients to txt file
-    rpcfile = common.tmpfile('.txt')
-    rpc.write_to_file(rpcfile)
-
-    if not os.path.exists(crop_colorized):
-        crop_colorized = ''
-    hij = " ".join(str(x) for x in H.flatten()) if H is not None else ""
-    command = ["colormesh", cloud, heights, rpcfile, crop_colorized, "-h", hij]
-    if ascii_ply:
-        command.append("--ascii")
-    if with_normals:
-        command.append("--with-normals")
-    if utm_zone:
-        command.extend(["--utm-zone", utm_zone])
-    if llbbx:
-        lonm, lonM, latm, latM = llbbx
-        command.extend(["--lon-m", lonm, "--lon-M", lonM, "--lat-m", latm, "--lat-M", latM])
-    if off_x:
-        command.extend(["--offset_x", "%d" % off_x])
-    if off_y:
-        command.extend(["--offset_y", "%d" % off_y])
-    common.run(command)
+    with rasterio.open(heights) as src:
+        h_map = src.read(1)
+
+    heights = h_map.ravel()
+    indices = np.indices(h_map.shape)
+
+    non_nan_ind = np.where(~np.isnan(heights))[0]
+
+    heights = heights[non_nan_ind]
+    cols = indices[1].ravel()[non_nan_ind]
+    rows = indices[0].ravel()[non_nan_ind]
+
+    if off_x or off_y:
+        cols = cols + (off_x or 0)
+        rows = rows + (off_y or 0)
+
+    # localize pixels
+    lons, lats = rpc.localization(cols, rows, heights)
+
+    # output CRS conversion
+    in_crs = geographiclib.pyproj_crs("epsg:4979")
+    pyproj_out_crs = geographiclib.pyproj_crs(cfg['out_crs'])
+    proj_com = "CRS {}".format(cfg['out_crs'])
+
+    if pyproj_out_crs != in_crs:
+        x, y, z = geographiclib.pyproj_transform(lons, lats,
+                                             in_crs, pyproj_out_crs, heights)
+        xyz_array = np.vstack((x, y, z)).T
+    else:
+        xyz_array = np.vstack((lons, lats, heights)).T
+
+    # write the point cloud to a ply file
+    if crop_colorized:
+        with rasterio.open(crop_colorized, 'r') as f:
+            img = f.read()
+        colors_list = img.transpose(1, 2, 0).reshape(-1, img.shape[0])[non_nan_ind]
+    else:
+        colors_list = None
+
+    ply.write_3d_point_cloud_to_ply(cloud, xyz_array, colors=colors_list,
+                                    extra_properties=None, extra_properties_names=None,
+                                    comments=["created by S2P",
+                                              "projection: {}".format(proj_com)])
```

Comment:
I feel like there is some mutual code between the present section (used in the more-than-2-images case), and [this section](https://github.com/lofrad/s2p/blob/7b4ff1c5a01d9614236f8294840b3e32a2cf11a4/s2p/__init__.py#L288-L309) (used in the 2-images case).

They could potentially be refactored into a single function called in both cases.
It would also solve the fact that with the current implementation (and the previous ones), 3d filtering was only done in the 2-images case, which is not very logical I believe.

Reply:
This was addressed by @lofrad in commit c9936f3b5e